File: 백준/Platinum/5541. 釘 （Nails）/釘 （Nails）.cc
#include <bits/stdc++.h>
using namespace std;

#define INF 1000000000000000000
#define MOD 1000000000

// Helper lambda functions
auto is_inbound = [](int pos_x, int size_x, int pos_y, int size_y) {
    return 0 <= pos_x && pos_x < size_x && 0 <= pos_y && pos_y < size_y;
};

// Delta for directions
const vector<pair<int, int>> DELTA = {{0, 0}, {0, -1}, {1, 0}, {0, 1}, {-1, 0}};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    // Input tokens
    int t = 1;
    vector<string> answers;

    for (int hh = 1; hh <= t; hh++) {
        int n, m;
        cin >> n >> m;

        // Triangle grid (padded with two extra rows/columns)
        vector<vector<int>> grid_triangle(n + 2, vector<int>(n + 2, 0));

        // Imos method for triangle: record on 6 cells and then accumulate
        for (int i = 0; i < m; i++) {
            int a, b, x;
            cin >> b >> a >> x;
            b -= 1; // Offset by 1 (as in the Python version)
            a -= 1; // Offset by 1
            x += 1;

            grid_triangle[b][a] += 1;
            grid_triangle[b][a + 1] -= 1;
            grid_triangle[b + x][a + x + 1] += 1;
            grid_triangle[b + x + 1][a + x + 1] -= 1;
            grid_triangle[b + x][a] -= 1;
            grid_triangle[b + x + 1][a + 1] += 1;

            // Debugging print for triangle grid (optional)
            // for (const auto& row : grid_triangle) {
            //     for (const auto& val : row) {
            //         cerr << val << " ";
            //     }
            //     cerr << "\n";
            // }
            // cerr << "\n";
        }

        // Left-right accumulation
        for (int y = 0; y < n + 2; y++) {
            int x = 1;
            while (is_inbound(x, n + 2, y, n + 2)) {
                grid_triangle[y][x] += grid_triangle[y][x - 1];
                x++;
            }
        }

        // Top-bottom accumulation
        for (int x = 0; x < n + 2; x++) {
            int y = 1;
            while (is_inbound(x, n + 2, y, n + 2)) {
                grid_triangle[y][x] += grid_triangle[y - 1][x];
                y++;
            }
        }

        // Diagonal accumulation
        for (int r = 0, dy_minus_dx = -n - 3; dy_minus_dx < n + 4; r++, dy_minus_dx++) {
            int x, y;
            if (dy_minus_dx < 0) {
                x = -dy_minus_dx;
                y = 0;
            } else {
                x = 0;
                y = dy_minus_dx;
            }
            x++;
            y++;
            while (is_inbound(x, n + 2, y, n + 2)) {
                grid_triangle[y][x] += grid_triangle[y - 1][x - 1];
                x++;
                y++;
            }
        }

        // Counting the final answer
        int answer = 0;
        for (const auto& row : grid_triangle) {
            for (const auto& val : row) {
                if (val) answer++;
            }
        }
        answers.push_back(to_string(answer));
    }

    // Output the answers
    for (const auto& ans : answers) {
        cout << ans << "\n";
    }

    return 0;
}

File: 백준/Platinum/5541. 釘 （Nails）/釘 （Nails）.py
from sys import setrecursionlimit, stdout, stderr

with open(0, 'r') as f:
    tokens = iter(f.read().split())
    input = lambda: next(tokens, None)
    print = lambda *args, sep="\n", end="\n": stdout.write(sep.join(map(str, args)) + end)
    eprint = lambda *args, sep=" ", end="\n": stderr.write(sep.join(map(str, args)) + end)
    fprint = lambda *args, sep=" ", end="\n", file: file.write(sep.join(map(str, args)) + end)
    is_inbound = lambda pos_x, size_x, pos_y, size_y: 0 <= pos_x < size_x and 0 <= pos_y < size_y
    DELTA = [(0, 0), (0, -1), (1, 0), (0, 1), (-1, 0)]
    INF = 10 ** 18
    MOD = 1_000_000_000
    t = 1
    answers = []
    for hh in range(1, t + 1):
        n, m = int(input()), int(input())
        # 삼각형용 그리드 (상하좌우로 두 칸씩 패딩 필요)
        grid_triangle = [[0 for c in range(n + 2)] for r in range(n + 2)]
        # imos법 사용
        # 삼각형은 6칸만 기록하고 좌우, 상하, 우하향 순으로 누적합
        for _ in range(m):
            b, a, x = (int(input()) + 1 for _ in range(3))
            b -= 2
            a -= 2
            grid_triangle[b][a] += 1
            grid_triangle[b][a + 1] -= 1
            grid_triangle[b + x][a + x + 1] += 1
            grid_triangle[b + x + 1][a + x + 1] -= 1
            grid_triangle[b + x][a] -= 1
            grid_triangle[b + x + 1][a + 1] += 1
        # 좌우
        for y in range(n + 2):
            x = 1
            while is_inbound(x, n + 2, y, n + 2):
                grid_triangle[y][x] += grid_triangle[y][x - 1]
                x += 1
        # 상하
        for x in range(n + 2):
            y = 1
            while is_inbound(x, n + 2, y, n + 2):
                grid_triangle[y][x] += grid_triangle[y - 1][x]
                y += 1
        # 우하향
        for r, dy_minus_dx in enumerate(range(-n - 3, n + 4)):
            if dy_minus_dx < 0:
                x, y = -dy_minus_dx, 0
            else:
                x, y = 0, dy_minus_dx
            x += 1
            y += 1
            while is_inbound(x, n + 2, y, n + 2):
                grid_triangle[y][x] += grid_triangle[y - 1][x - 1]
                x += 1
                y += 1
        answer = sum(sum(bool(v2) for v2 in v) for v in grid_triangle)
        answers.append(f"{answer}")
print(*answers, sep="\n")
